### DISCUSSION TITLE
Image reconstruction and raster plots in eth_mnist.py

### DISCUSSION THREAD

**Discussion by rafaelblevin821** [BindsNET/bindsnet#581] [General]
Hi, so just looking at the samples below I have a few questions...

![eth_mnist1](https://user-images.githubusercontent.com/1954873/188684068-4e0f3c4e-e59b-4ae0-aff1-012880e96da7.png)
![eth_mnist2](https://user-images.githubusercontent.com/1954873/188684070-6acbb025-b98f-40f8-ad95-fda7df9739d3.png)
![eth_mnist3](https://user-images.githubusercontent.com/1954873/188684071-6e88d6bb-d291-4207-98ef-dcc477290ee3.png)

Referring to image reconstruction, what is the point of the image reconstruction? Is this the Poisson reconstructed image before it is passed through the network or is this after it has been passed through the network?

The other question is, how do the spikes in the raster plot represent the digits? The spikes are very sparse, and the learning recognition is not easily conceptualised from the raster plot of spikes, due to so little information.

Any help would be greatly appreciated. Thank you.

**Comment by SimonInParis** [comment]
The image reconstructs the MNIST from the Poisson conversion, as the the sum of spikes for every 28x28 pixel occuring during the model running time, that is passed to the input of the model:
`inputs["X"].view(time, 784).sum(0).view(28, 28)`

Note: you can try to lower the running time param to look at coarser reconstructions of the original MNIST images, it will also slightly affect the model performance:
`parser.add_argument("--time", type=int, default=250)`

The bottom graph ('X') shows the 784 (28x28) Poisson converted pixels, during the 250 time steps. The denser the dots in time, the higher the brightness of the pixel.

The 'Ae' graph shows which excitatory output neurons reacts the most to the input Poisson stream. These are the neurons used to predict the actual MNIST class, through the supervised `all_activity` mechanism, that adaptively assigns classes to the most frequent activity of each output neuron:
Here is the regular update of the output neurons assignment - the _weakly supervised_ part of the model:
```
        # Add to spikes recording.
        spike_record[step % update_interval] = spikes["Ae"].get("s").squeeze()
```
            # Assign labels to excitatory layer neurons.
            assignments, proportions, rates = assign_labels(
                spikes=spike_record,
                labels=label_tensor,
                n_labels=n_classes,
                rates=rates,
            )

and here is the class prediction, estimated by the assignments of the output neurons activity:
```
            all_activity_pred = all_activity(
                spikes=spike_record, assignments=assignments, n_labels=n_classes
            )
```
```
            # Compute network accuracy according to available classification strategies.
            accuracy["all"].append(
                100
                * torch.sum(label_tensor.long() == all_activity_pred).item()
                / len(label_tensor)
            )
```
(`all_activity` and `proportion_weighting` are somewhat different assignment strategies, but show very similar resulting accuracies).
As for now, I don't think we can display the output neurons assignment map. Maybe you can do it?

The 'Ai' shows the inhibitory neurons spikes. These spikes prevent other excitatory output neurons from spiking, and are part of the ETH model design (from [models.py](https://github.com/BindsNET/bindsnet/blob/61fda794534bf3c85307f8f71a6124ff5c7c4b52/bindsnet/models/models.py#L87)):
```
        # Add to network
        self.add_layer(input_layer, name="X")
        self.add_layer(exc_layer, name="Ae")
        self.add_layer(inh_layer, name="Ai")
        self.add_connection(input_exc_conn, source="X", target="Ae")
        self.add_connection(exc_inh_conn, source="Ae", target="Ai")
        self.add_connection(inh_exc_conn, source="Ai", target="Ae")
```

**Reply by rafaelblevin821** [reply]
Hi @SimonInParis , thank you for taking the time to reply; much appreciated.

I wonder if you would just clarify a few things:
> The image reconstructs the MNIST from the Poisson conversion, as the the sum of spikes for every 28x28 pixel occuring during the model running time, that is passed to the input of the model:
> `inputs["X"].view(time, 784).sum(0).view(28, 28)`

So, you are saying that the first image is the Poisson encoded image and the second image (reconstructed image) has been passed through the model and reconstructed after output?

Also:
> The bottom graph ('X') shows the 784 (28x28) Poisson converted pixels, during the 250 time steps. The denser the dots in time, the higher the brightness of the pixel.
> 
> The 'Ae' graph shows which excitatory output neurons react the most to the input Poisson stream. These are the neurons used to predict the actual MNIST class through the supervised `all_activity` mechanism, that adaptively assigns classes to the most frequent activity of each output neuron:

So the ('X') plot is the Poisson encoded input? But how does it determine the label from the Ae/Ai spiking activity? For example, in the last image I posted, which was labeled 6, there are around 30 spikes within the 250ms stimulation time, and they seem to be at regular intervals. How do these 30 spikes predict the label of '6'?

Thank you, I appreciate your help on this.

**Comment by Hananel-Hazan** [comment]
> So, you are saying that the first image is the Poisson encoded image and the second image (reconstructed image) has been passed through the model and reconstructed after output?

Yes

> How do these 30 spikes predict the label of '6'?

All the details is in the accompany papers, please read [Diehl and Cook](https://www.frontiersin.org/articles/10.3389/fncom.2015.00099/full) paper or [Unsupervised learning of digit recognition using spike-timing-dependent plasticity](https://www.frontiersin.org/article/10.3389/fncom.2015.00099) and for more details [Lattice map spiking neural networks (LM-SNNs) for clustering and classifying image data](https://link.springer.com/article/10.1007/s10472-019-09665-3) or [arXiv](https://arxiv.org/pdf/1906.11826).
In short, the image you see here is reconstructed from the weight matrix of the neuron that spike the most. The mechanization on how the algorithm assign the label was describe beautifully by @SimonInParis above and for more details please read the above papers, all the explanations are there.

**Reply by rafaelblevin821** [reply]
> > So, you are saying that the first image is the Poisson encoded image and the second image (reconstructed image) has been passed through the model and reconstructed after output?
> 
> Yes
> 
> > How do these 30 spikes predict the label of '6'?
> 
> All the details is in the accompany papers, please read [Diehl and Cook](https://www.frontiersin.org/articles/10.3389/fncom.2015.00099/full) paper or [Unsupervised learning of digit recognition using spike-timing-dependent plasticity](https://www.frontiersin.org/article/10.3389/fncom.2015.00099) and for more details [Lattice map spiking neural networks (LM-SNNs) for clustering and classifying image data](https://link.springer.com/article/10.1007/s10472-019-09665-3) or [arXiv](https://arxiv.org/pdf/1906.11826). In short, the image you see here is reconstructed from the weight matrix of the neuron that spike the most. The mechanization on how the algorithm assign the label was describe beautifully by @SimonInParis above and for more details please read the above papers, all the explanations are there.

Hi @Hananel-Hazan and @SimonInParis , 

I am very grateful for the assistance in understanding how this network works. I had a problem conceptualizing some aspects of the learning and classification methods. I have already read the Diehl & Cook paper and the Unsupervised learning of digit recognition using STDP paper. It was just some small details that I needed to be clarified, but you have both been very helpful. Thank you.
